control_plane/src/background_process.rs

```diff
@@ -0,0 +1,250 @@
+//! Spawns and kills background processes that are needed by Neon CLI.
+//! Applies common set-up such as log and pid files to every process.
+//! Neon CLI does not run in background, so it needs to store the information about
+//! spawned processes which it does in this module.
+
+use std::ffi::OsStr;
+use std::io::Write;
+use std::path::Path;
+use std::process::{Child, Command};
+use std::time::Duration;
+use std::{fs, io, thread};
+
+use anyhow::{anyhow, bail, Context, Result};
+use nix::errno::Errno;
+use nix::sys::signal::{kill, Signal};
+use nix::unistd::Pid;
+
+use utils::lock_file;
+
+const RETRIES: u32 = 15;
+const RETRY_TIMEOUT_MILLIS: u64 = 500;
+
+/// Ensures process startup handles pidfile, either by creating the new one
+/// or by checking the one that the spawned process should create.
+pub enum InitialPidFile<'t> {
+    /// Creates a special pidfile, to allow future CLI invocations to manipulate the process.
+    Create(&'t Path),
+    /// A process is creating the pidfile itself, need to wait for that event.
+    Expect(&'t Path),
+}
```

Comment:
The process itself should always create it.

Reply:
The problem is that `etcd` does not create one, but it's the only way our CLI can track that something was launched between restarts.
So for etcd and any other lock-less processes, we would still want to create some PID file and do the same stdout/stderr binding, same as for our storage processes that do create their PID files themselves.

Reply:
I see. I tested launching two etcd instances in the same data directory:

```
2022-10-30 11:28:56.990044 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2022-10-30 11:28:56.992037 C | etcdmain: listen tcp 127.0.0.1:2380: bind: address already in use
```
So that's the first line of defence against having two etcd processes. I then tried it with a different port (`etcd --listen-peer-urls="http://127.0.0.1:1234" --listen-client-urls="http://127.0.0.1:1235" --advertise-client-urls="http://127.0.0.1:1235"`):

```
2022-10-30 11:30:19.253997 W | etcdserver: another etcd process is using "default.etcd/member/snap/db" and holds the file lock, or loading backend file is taking >10 seconds
2022-10-30 11:30:19.254135 W | etcdserver: waiting for it to exit before starting...
```
So etcd uses a file-lock on `default.etcd/member/snap/db`, but it doesn't create a PID file.

Reply:
I've left it so far, since `etcd` still needs its pid persisted somewhere on FS (with the current approach)